ext/semian/sysv_semaphores.c

```diff
@@ -107,11 +107,9 @@ perform_semop(int sem_id, short index, short op, short flags, struct timespec *t
   buf.sem_op  = op;
   buf.sem_flg = flags;
 
-  int num_retries = 3;
-
   do {
     result = semtimedop(sem_id, &buf, 1, ts);
-  } while (result < 0 && errno == EINTR && num_retries-- > 0);
+  } while (result < 0 && errno == EINTR);
```

Comment:
`semtimedop` when interrupted by signal, does not get restarted automatically, Initially in https://github.com/Shopify/semian/pull/233
 I set the retry count to `3` but that only (partially) fixed `Semian::InternalError: error setting ticket count, errno: 4 (Interrupted system call)` in CI (VM), it was still happening but a lot less

Running it locally in docker would work without any errors but running docker in CI would again result in system calls being interrupted.

Reply:
in theory it would be fine? unless each syscall is constantly getting interrupted

Reply:
As we discussed it still scares me to introduce the possibility of an infinite loop in hard-to-debug C extensions. If there needs to be a higher limit, we should just increase the limit IMO.

I'd very much like to know the source of these signals, given the high number of occurrences. In GUI environments, a common cause is SIGWINCH from resizing a window, however that's not possible here. It could be a profiler attached that's using SIGALRM on a timer, but we don't run stackprof in this project. Could attach a signal handler to every signal and log.

Reply:
I guess it should be fine, since I can't think of a situation where result would forever be less than 0 with errno also remaining `EINTR`.

That being said, I have two issues here. First, we seem to be solving a CI problem, not a production problem. Second, but related to the first, is that this feels like the wrong place to solve this. If there were a signal, we completely suppress it. Perhaps the application would want to handle the signal differently? I feel like this logic should exist at the Ruby level with the ability to be configured by the application that's using Semian. We can default it to "forever retry". Something like this, for example:

```ruby
Semian.interrupt_handler = -> { |signal| Signal.signame(signal) != "TERM" }
```

If the result of the block is `true`, it retries the `semtimedop` call, otherwise it fails. Another option would be to have something like `Semian.interrupt_retry_count`. We could set it to 3 for production, and some really high value for CI.

What does everyone else think?

Reply:
I'm OK leaving it at 3 for prod/CI.
If CI requires a much larger number, then we should revisit (exponential backoff?).

Reply:
Retrying on EINTR is correct and you could consider it a bug that we were not doing so. This does not suppress any signal handlers. EINTR is an implementation detail of signals. This is my understanding:

1. Thread gets a signal, but previously issued a blocking syscall and is currently blocked.
2. Signal handlers don't get their own parallel threads. The syscall needs to exit to unblock and let the signal handler run. It returns EINTR.
3. The signal handler is run instead of returning to the syscall caller.
4. After the signal handler returns, control flow resumes at the syscall. The program can choose to retry.

However, we should really just be retrying up until the timeout. Retries == 3 is probably not going to blow any time budgets, but if we increase it much further, it starts to become a concern.

Reply:
https://www.gnu.org/software/libc/manual/html_node/Interrupted-Primitives.html

Reply:
For the record, that's:

```c
# define TEMP_FAILURE_RETRY(expression) \
(__extension__                                                              \
({ long int __result;                                                      \
do __result = (long int) (expression);                                      \
while (__result == -1L && errno == EINTR);                              \
__result; }))
```

Which is guarded by `__USE_GNU`. I recently made a commit that switched to `gnu99` format, so we should be OK to use `TEMP_FAILURE_RETRY` if you want (although it's unbounded).

Reply:
Some other thinking I didn't include above:

When handling a signal, an application can either cancel execution, or let it continue. If it cancels it (with `exit` or whatever), it doesn't matter that we would have retried, as the process will run atexit handlers and die before we can. If it doesn't cancel it, the signal should not raise an error in whatever random syscall was being made at the time. That would break the correctness of the program.

However, a common pattern is to handle a signal, and mark a flag to gracefully terminate. In this case, we still need to make sure we run all logic correctly. In semian's case, this is really important to avoid corrupting semaphore state. However, we don't want to block forever. That's one reason why it's important to respect the timeout. I wonder if we should remove the null timeout path entirely 🤔

Reply:
> However, we should really just be retrying up until the timeout

This seems like the best option, since that can be configured at the application level. The only function that doesn't use a timed operation is [`sem_meta_unlock `](https://github.com/Shopify/semian/blob/fc4f4e9d95215e3d45e61730adea0bff1913ddb2/ext/semian/sysv_semaphores.c#L141).

Reply:
The unlock path shouldn't ever actually block, maybe let's just give it a long timeout and require some timeout is passed.

Reply:
> TEMP_FAILURE_RETRY

Yeah, let's not use that as (like Michael pointed out) it's unbounded, but I quite like all the context provided on that doc page.

Reply:
I reverted this change in this PR.